### ISSUE TITLE
0.16.3: composer-client always shows database update page

### ISSUE THREAD

**Issue by fritzmg** [contao-community-alliance/composer-client#279]
Since `0.16.3` the `composer-client` now _always_ shows the database update page, even if there is nothing to be done (i.e. the page is empty).

![screen-shot-2016-02-19-at-10 56 46](https://cloud.githubusercontent.com/assets/4970961/13172288/aaca9800-d6f7-11e5-987f-dd555ce6b560.png)

Previously the `composer-client` would skip that page, if no actual database update was necessary.

**Comment by backbone87** [comment]
from what i know, this could happen if you use extension with runonce files? and since on every composer update the runonce files gets executed, even if no new packages were installed, you get to the db upgrade screen

**Comment by discordier** [comment]
Please check the log if there has been a runonce file executed.

**Comment by discordier** [comment]
Nope, not on mine.

**Comment by fritzmg** [comment]
Ok, I have found the bug. Once you install an extension which causes a database update and you let the database update and then run a package update _again_ (or go into the database update screen manually, doesn't matter), it will show the empty database update page.

After
```php
$installer = \System::importStatic('Database\Installer');
$form = $installer->generateSqlForm();
```
in the `UpdateDatabaseController`, `$_SESSION['sql_commands']` and `$form` will have the following content:

`$_SESSION['sql_commands']`:
```
array(3) {
  ["8c0b1c617d59c4bd7db06a571d1b7f63"]=>
  string(40) "ALTER TABLE `tl_article` DROP `inherit`;"
  ["b31e2f9342aae5a6d8b600e1b8b0d771"]=>
  string(45) "ALTER TABLE `tl_article` DROP `inheritLevel`;"
  ["08b16265d57141031a7aff3b63040cfd"]=>
  string(48) "ALTER TABLE `tl_article` DROP `inheritPriority`;"
}
```

`$form`:
```
string(0) ""
```

The problem is, that the array in `$_SESSION['sql_commands']` never actually gets cleared after the first database update.

There are two places, where `$_SESSION['sql_commands']` is cleared:

- Once in [UpdateDatabaseController.php#L50](https://github.com/contao-community-alliance/composer-client/blob/0.16.3/src/system/modules/!composer/src/Controller/UpdateDatabaseController.php#L50)
- Once in [Contao/Database/Installer.php#L48](https://github.com/contao/core/blob/3.5.6/system/modules/core/library/Contao/Database/Installer.php#L48)

The first one is only reached, when the database update form is submitted and, more importantly, the second one from Contao itself is only reached, when there are actually SQL commands to be executed. So when there is no database update to be done, `$_SESSION['sql_commands']` does not get cleared by `$installer->generateSqlForm();` and the following redirect:
```php
if (empty($_SESSION['sql_commands'])) {
    Messages::addInfo($GLOBALS['TL_LANG']['composer_client']['databaseUptodate']);
    $this->redirect('contao/main.php?do=composer');
}
```
is __not__ executed.

I actually saw the empty database update page _every time_ because I never actually submitted the empty database form (because, what's the point, I thought ;)).

**Comment by fritzmg** [comment]
To fix it, either always clear `$_SESSION['sql_commands']` before `$form = $installer->generateSqlForm();` or change
```php
if (empty($_SESSION['sql_commands'])) {
```
to
```php
if (!$form) {
```
for example.

**Comment by discordier** [comment]
Ok, we will have to fix this then.
I'd rather go with clearing the session prior generating the form as it will get filled again within there.

Makes most sense to me.

**Comment by discordier** [comment]
Should be fixed with 816d15c. @fritzmg could you give it a test run?

**Comment by fritzmg** [comment]
Yes, seems to work fine.
